apollo-router/src/protocols/websocket.rs

```diff
@@ -361,7 +388,129 @@ where
         })
 }
 
-impl<S> Stream for GraphqlWebSocket<S>
+pub(crate) struct SubscriptionStream<S> {
+    inner_stream: SplitStream<InnerStream<S>>,
+    close_signal: Option<tokio::sync::oneshot::Sender<()>>,
+}
+
+impl<S> SubscriptionStream<S>
+where
+    S: Stream<Item = serde_json::Result<ServerMessage>>
+        + Sink<ClientMessage>
+        + std::marker::Unpin
+        + std::marker::Send
+        + 'static,
+{
+    pub(crate) fn new(
+        stream: S,
+        id: String,
+        protocol: WebSocketProtocol,
+        heartbeat_interval: Option<tokio::time::Duration>,
+    ) -> Self {
+        let (mut sink, inner_stream) = InnerStream::new(stream, id, protocol).split();
+        let (close_signal, close_sentinel) = tokio::sync::oneshot::channel::<()>();
+
+        tokio::task::spawn(async move {
+            if let (WebSocketProtocol::GraphqlWs, Some(duration)) = (protocol, heartbeat_interval) {
+                let mut interval =
+                    tokio::time::interval_at(tokio::time::Instant::now() + duration, duration);
+                interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+                let mut heartbeat_stream = IntervalStream::new(interval)
+                    .map(|_| {
+                        Ok(ClientMessage::Ping {
+                            payload: Some(serde_json_bytes::Value::String(
+                                "APOLLO_ROUTER_HEARTBEAT".into(),
+                            )),
+                        })
```

Comment:
this is fine to leave as is for now, but as a follow up, we should actually check if we got a response to the heartbeat. Right now it makes sure that the websocket connection stays open, but it does not account for the possibility of the backend hanging